Improve safe transmute error reporting

bryangarza opened the pull request:
This patch updates the error reporting when Safe Transmute is not possible between 2 types by including the reason.

Also, fix some small bugs that occur when computing the `Answer` for transmutability.

bryangarza commented:
cc @jswrenn

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4571057596/jobs/8069003574) [(plain)](https://github.com/rust-lang/rust/commit/f9b7248e7cc0f495eccdb8d40d0d580adbd1abf0/checks/12414693892/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:8f4b7f84864484a7bf31766abe9204da3cbe65b3)
Download action repository 'rust-lang/simpleinfra@master' (SHA:3040325909b538d8ad81ad89a04b7a91b109c313)
Complete job name: PR (x86_64-gnu-llvm-14, false, ubuntu-20.04-16core-64gb)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-llvm-14
---
diff of stderr:

2   --> $DIR/transmute-padding-ice.rs:27:40
3    |
4 LL |     assert::is_maybe_transmutable::<B, A>();
-    |                                        ^ `B` cannot be safely transmuted into `A` in the defining scope of `assert::Context`.
+    |                                        |
+    |                                        |
+    |                                        `B` cannot be safely transmuted into `A` in the defining scope of `assert::Context`.
+    |                                        Reason: DstIsTooBig
6    |
7    = help: the trait `BikeshedIntrinsicFrom<B, assert::Context, Assume { alignment: true, lifetimes: true, safety: true, validity: true }>` is not implemented for `A`
8 note: required by a bound in `is_maybe_transmutable`

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/transmute/transmute-padding-ice/transmute-padding-ice.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args transmute/transmute-padding-ice.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/transmute/transmute-padding-ice.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/transmute/transmute-padding-ice" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/transmute/transmute-padding-ice/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0277]: `B` cannot be safely transmuted into `A` in the defining scope of `assert::Context`.
  --> fake-test-src-base/transmute/transmute-padding-ice.rs:27:40
   |
LL |     assert::is_maybe_transmutable::<B, A>();
   |                                        |
   |                                        |
   |                                        `B` cannot be safely transmuted into `A` in the defining scope of `assert::Context`.
   |                                        Reason: DstIsTooBig
   |
   = help: the trait `BikeshedIntrinsicFrom<B, assert::Context, Assume { alignment: true, lifetimes: true, safety: true, validity: true }>` is not implemented for `A`
note: required by a bound in `is_maybe_transmutable`
  --> fake-test-src-base/transmute/transmute-padding-ice.rs:11:14
   |
LL |       pub fn is_maybe_transmutable<Src, Dst>()
LL |       where
LL |       where
LL |           Dst: BikeshedIntrinsicFrom<
LL | |             Src,
LL | |             Context,
LL | |             Context,
LL | |             { Assume { alignment: true, lifetimes: true, safety: true, validity: true } },
LL | |         >,
   | |_________^ required by this bound in `is_maybe_transmutable`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
------------------------------------------
```

</details>

bryangarza commented:
I think I missed some tests that need to be blessed, I’ll do it tomorrow

jswrenn commented on compiler/rustc_transmute/Cargo.toml:
```diff
@@ -14,6 +14,7 @@ rustc_macros = { path = "../rustc_macros", optional = true}
 rustc_middle = { path = "../rustc_middle", optional = true}
 rustc_span = { path = "../rustc_span", optional = true}
 rustc_target = { path = "../rustc_target", optional = true}
+rustc_transmute_types = { path = "../rustc_transmute_types", optional = true}
```
I don't think this can be optional. To check, `cd` into `rustc_transmute` and run `cargo check`.

bryangarza replied:
thanks. In my latest commit I fixed the build (had to gate some more code behind the rustc flag because we depend on rustc_middle more)

jswrenn commented on tests/ui/transmutability/enums/repr/should_require_well_defined_layout.stderr:
```diff
@@ -2,7 +2,10 @@ error[E0277]: `void::repr_rust` cannot be safely transmuted into `()` in the def
+   |                                                    ^^
+   |                                                    |
+   |                                                    `void::repr_rust` cannot be safely transmuted into `()` in the defining scope of `assert::Context`.
+   |                                                    Reason: SrcIsUnspecified
```
Instead of:
```
`void::repr_rust` cannot be safely transmuted into `()` in the defining scope of `assert::Context`.
Reason: SrcIsUnspecified
```
I'd suggest:
```
`void::repr_rust` cannot be safely transmuted into `()` in the defining scope of `assert::Context` because `void::repr_rust` does not have a well-specified layout.
```

bryangarza replied:
Is it possible to combine them somehow? The first line comes from the on_unimplemented macro and the other line comes from the actual error.

compiler-errors replied:
It's probably better to just remove the `rustc_on_unimplemented` from the macro and handle formatting the error message for the transmute trait manually.

jswrenn commented on tests/ui/transmutability/enums/should_pad_variants.stderr:
```diff
@@ -2,7 +2,10 @@ error[E0277]: `Src` cannot be safely transmuted into `Dst` in the defining scope
+   |                                    ^^^
+   |                                    |
+   |                                    `Src` cannot be safely transmuted into `Dst` in the defining scope of `should_pad_variants::Context`.
+   |                                    Reason: DstIsTooBig
```
I'd suggest:
```
`Src` cannot be safely transmuted into `Dst` in the defining scope of `should_pad_variants::Context` because the size of `Src` is smaller than the size of `Dst` (YYY bytes).
```

jswrenn commented on tests/ui/transmutability/enums/should_respect_endianness.stderr:
```diff
@@ -2,7 +2,10 @@ error[E0277]: `Src` cannot be safely transmuted into `Unexpected` in the definin
+   |                                    ^^^^^^^^^^
+   |                                    |
+   |                                    `Src` cannot be safely transmuted into `Unexpected` in the defining scope of `assert::Context`.
+   |                                    Reason: DstIsBitIncompatible
```
I'd suggest:
```
`Src` cannot be safely transmuted into `Unexpected` in the defining scope of `assert::Context` because at least one value of `Src` isn't a bit-valid value of `Unexpected`.
```

Eventually, we'll want to add more information here about why (namely, by giving an example value of the source type that's not a valid instance of the destination type), but we can punt that to future work.

jswrenn commented on tests/ui/transmutability/visibility/should_reject_if_dst_has_private_field.stderr:
```diff
@@ -2,7 +2,10 @@ error[E0277]: `Src` cannot be safely transmuted into `Dst` in the defining scope
-   |                                         ^^^^^^^^ `Src` cannot be safely transmuted into `Dst` in the defining scope of `test::Context`.
+   |                                         ^^^^^^^^
+   |                                         |
+   |                                         `Src` cannot be safely transmuted into `Dst` in the defining scope of `test::Context`.
+   |                                         Reason: DstIsPrivate
```
I'd suggest:
```
`Src` cannot be safely transmuted into `Dst` in the defining scope of `test::Context` because `Dst` is or contains a type or field that is not visible in that scope.
```

compiler-errors commented on compiler/rustc_transmute_types/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
```
I don't think we really need to put this into a separate crate. Most type definitions like this just live in rustc_middle.

compiler-errors commented:
use `@rustbot ready` to mark this as ready to review

@rustbot author

bryangarza committed:
Improve safe transmute error reporting

This patch updates the error reporting when Safe Transmute is not
possible between 2 types by including the reason.

Also, fix some small bugs that occur when computing the `Answer` for
transmutability.

bryangarza commented on compiler/rustc_middle/src/traits/mod.rs:
```diff
@@ -569,10 +570,21 @@ pub struct DerivedObligationCause<'tcx> {
 pub enum SelectionError<'tcx> {
     /// The trait is not implemented.
-    Unimplemented,
+    Unimplemented(Option<UnimplementedReason>),
```
Thanks, I created a new variant (`rustc_transmute` does a lot of work, so don't want to run it twice)

bryangarza commented on compiler/rustc_middle/src/traits/mod.rs:
```diff
@@ -590,6 +591,12 @@ pub enum SelectionError<'tcx> {
     /// Signaling that an error has already been emitted, to avoid
     /// multiple errors being shown.
     ErrorReporting,
+    NotTransmutable {
+        reason: NonTransmutabilityReason,
```
I named the variant `NotTransmutable` because the other variants also use `Not`, and I named the reason type `NonTransmutabilityReason` because `NotTransmutableReason` reads kind of weird IMO, but I can rename them to have the same prefix if we want.

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:
```diff
@@ -616,7 +617,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
 
         let mut err = match *error {
-            SelectionError::Unimplemented => {
+            SelectionError::Unimplemented | SelectionError::NotTransmutable { .. } => {
```
I would probably just handle this branch separately.

compiler-errors replied:
Actually, see comment below. I'm not convinced this needs a separate `SelectionError`.

bryangarza replied:
I had it separate but then I realized I would have to add a bunch of the code from this one in order to create the full error (for example, including the code block and pointing to say "required by a bound..."...

compiler-errors replied:
Hm, this function is getting somewhat out of control. I wonder if you could factor it out in some way to make it more manageable...

Anyways, if you go ahead with my advice below and just use `Unimplemented` then it's kind of moot, but really this could use some investment in clean-up if you're up for it.

bryangarza replied:
I kept a separate variant, and was able to show the rest of the original error message by only copying over the necessary logic from the `Unspecified` case. But I can put up another PR that refactors the `Unspecified` case so we can reuse pieces and have it not be all in one giant code block.

compiler-errors commented on compiler/rustc_trait_selection/src/traits/select/confirmation.rs:
```diff
@@ -302,6 +302,12 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+            Answer::No(reason) => Err(SelectionError::NotTransmutable {
+                reason,
+                scope,
+                src: type_at(1).skip_binder(),
```
Do you need to return these types? I think you can extract them from the `FulfillmentError` during error reporting?

compiler-errors replied:
Also, does this even need to be a separate `SelectionError`? I think you can just check if the [`obligation`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/traits/struct.FulfillmentError.html#structfield.obligation) in the `FulfillmentError` is the transmutability trait, and extract the info again?

bryangarza replied:
is that the case? I don't understand how to get these fields from `PredicateObligation` -- do you have an example?

compiler-errors replied:
Well you should have access to the trait predicate - https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_trait_selection/traits/error_reporting/mod.rs.html#653

Then from the trait predicate, you'd just be extracting the types from the substitutions just like the transmute trait confirmation code does.

bryangarza replied:
oh, gotcha, yeah that makes more sense now. I missed that `let trait_predicate = ...` line before.. will try it out

bryangarza replied:
Okay, I got it to work. I kept a separate `SelectionError` so that I can include the specific `NotTransmutable` error -- I don't think that is possible to have access to without either having the separate variant or calling rustc_transmute again.

bryangarza commented:
FYI
- easier to review if you hide whitespace changes
- All the changes in `rustc_transmute` are just moving some definitions to be under `mod rustc` -- no actual code changes there

compiler-errors commented on tests/ui/transmutability/arrays/should_require_well_defined_layout.stderr:
```diff
@@ -1,10 +1,9 @@
    |
 LL |         assert::is_maybe_transmutable::<repr_rust, ()>();
-   |                                                    ^^ `[String; 0]` cannot be safely transmuted into `()` in the defining scope of `assert::Context`.
+   |                                                    ^^ the trait `BikeshedIntrinsicFrom<[String; 0], assert::Context, Assume { alignment: true, lifetimes: true, safety: true, validity: true }>` is not implemented for `()`
```
This should probably be reverted back to what it used to be.

compiler-errors replied:
Or at least made better, this current rendering is pretty rough.

bryangarza commented:
@rustbot ready

compiler-errors commented:
Some more comments. Other than those, I'm pretty happy with the state of this PR, though.

@rustbot author

bryangarza commented on compiler/rustc_transmute/src/layout/tree.rs:
```diff
@@ -195,7 +195,7 @@ pub(crate) mod rustc {
         fn from(err: LayoutError<'tcx>) -> Self {
             match err {
-                LayoutError::Unknown(..) => Self::Unknown,
+                LayoutError::Unknown(..) => Self::TypeError,
```
I'll add back the `Unknown` case

bors commented:
:umbrella: The latest upstream changes (presumably #110214) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented:
r=me after rebasing

there may be some problems with the error reporting with the new trait solver, since we *will* need to recompute the cannot-transmute-reason on demand in the error reporting code...

you can probably just leave that though, I can figure it out, unless you want to. ping me if you need help.

bryangarza commented on compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:
```diff
@@ -663,6 +700,41 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         }
                         let trait_ref = trait_predicate.to_poly_trait_ref();
+
+                        // FIXME(bryangarza): Move this logic to its own function
+                        // When using the new trait solver, recompute the safe transmute reason and use that for the error.
+                        if Some(trait_ref.def_id()) == self.tcx.lang_items().transmute_trait()
+                            && self.tcx.trait_solver_next()
+                        {
+                            let cause = obligation.cause.clone();
+                            let src_and_dst =
+                                trait_predicate.map_bound(|p| rustc_transmute::Types {
+                                    dst: p.trait_ref.substs.type_at(0),
+                                    src: p.trait_ref.substs.type_at(1),
+                                });
+                            let scope = trait_ref.skip_binder().substs.type_at(2);
+                            let Some(assume) =
+                                rustc_transmute::Assume::from_const(self.infcx.tcx, obligation.param_env, trait_ref.skip_binder().substs.const_at(3)) else {
+                                    bug!("Unable to construct rustc_transmute::Assume where it was previously possible.");
+                                };
+                            match rustc_transmute::TransmuteTypeEnv::new(self.infcx)
+                                .is_transmutable(cause, src_and_dst, scope, assume)
+                            {
+                                // Should never get a Yes at this point! We already ran it before, and did not get a Yes.
+                                rustc_transmute::Answer::Yes => bug!(
+                                    "Inconsistent rustc_transmute::is_transmutable(...) result."
+                                ),
+                                rustc_transmute::Answer::No(reason) => {
+                                    return self.report_selection_error(
+                                        obligation,
+                                        root_obligation,
+                                        &SelectionError::NotTransmutable(reason),
+                                    );
+                                }
+                                _ => (),
+                            };
+                        }
```
@compiler-errors Added this so the new trait solver can emit the right error, is this what you had in mind?

compiler-errors replied:
I think I asked this before, but why do we even need a special selection error kind? You have shown that it's possible to recompute the transmutability reason on the fly here :)

bryangarza replied:
It was just to avoid computing the safe transmute reason twice, but if the new trait solver is going to need it 2x anyways, we could just remove `SelectionError::NotTransmutable` variant

compiler-errors commented:
Thanks @bryangarza for dealing with many rounds of review.

@bors r+

bors commented:
:pushpin: Commit 6a4b2e89ff25f85507f603a4724f1ddadc16a1c9 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 6a4b2e89ff25f85507f603a4724f1ddadc16a1c9 -->
<!-- homu: {"type":"Approved","sha":"6a4b2e89ff25f85507f603a4724f1ddadc16a1c9","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bryangarza commented:
@compiler-errors I force pushed a small change after approval (to get `rustc_transmute` to successfully pass `cargo test`) -- is it possible to get this new commit hash to be the one in the queue?

compiler-errors commented:
Force-pushing automatically un-approves your PR. I will re-approve it when CI is green.

bryangarza commented:
@rustbot ready

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 36febe1f4da6ea7e066983786171d5f85f7473d9 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 36febe1f4da6ea7e066983786171d5f85f7473d9 -->
<!-- homu: {"type":"Approved","sha":"36febe1f4da6ea7e066983786171d5f85f7473d9","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
